shader/svgf_atrous.comp

```diff
@@ -1,185 +1,140 @@
-                w *= max(exp(-dist_l * dist_l * sigma_l), 0.0001f); // luminance based weight
-                // w *= kernel[kernel_index];
-                w *= wavelet_kernel[abs(xx)][abs(yy)];
-                ++kernel_index;
-
-                if (control.level < 2)
+                const vec4 color_q = (control.level & 1) == 0 ? imageLoad(diffuse_ping, q) : imageLoad(diffuse_pong, q);
+                const vec4 spec_q =  (control.level & 1) == 0 ? imageLoad(specular_ping, q) : imageLoad(specular_pong, q);
+                const float lum_q = rgb_to_luminance(color_q.rgb);
+                const float z_q = imageLoad(in_linear_depth, q).r;
+                const vec3 n_q = unpack_gbuffer_normal(imageLoad(in_normal, q).rg);
+                const float roughness_q = imageLoad(in_material, q).y;
+
+                float w_n = pow(max(0.0, dot(n_q, n_center)), control.sigma_n);
+                float w_z = abs(z_q - z_center.x) / (control.sigma_z * abs(dot(depth_grad, p.xy - q.xy)) + 1e-6);
+                float w_l = abs(lum_q - center_luminance) / phi_color;
+                float w = w_n * exp(-max(0.0, w_z) - max(0.0, w_l));
+
+                float r_max = 0.1;
+                float r_near = 0.01;
+                float r_far = 0.1;
+                float d_near = 10.0 * roughness_center;
+                float d_far = 70.0 * roughness_center;
+                float w_r = 1.0 - smoothstep(r_near, r_far, abs(roughness_q - roughness_center));
+                float w_d = 1.0 - smoothstep(d_near, d_far, length(vec3(p - q)));
+
+                float w_spec = exp(-max(0.0, w_z)) * w_n * w_r * w_d;
+                sum_w += w;
+                sum_color += color_q * vec4(vec3(w), w * w);
+                if (control.level < control.spec_iteration_count)
                 {
-                    sum_specular += specular_o.rgb * w;
-                    sum_w_spec += w;
+                    sum_w_spec += w_spec;
+                    sum_specular += spec_q * vec4(vec3(w_spec), w_spec * w_spec);
                 }
-                sum_color += color_o.rgb * w;
-                sum_variance += color_o.a * w * w;
-                sum_w_diff += w;
             }
         }
 
-        sum_color /= sum_w_diff;
-        sum_specular /= sum_w_spec;
-        sum_variance /= sum_w_diff * sum_w_diff;
-
-        sum_color = max(vec3(0.0), sum_color);
-        sum_specular = max(vec3(0.0), sum_specular);
-
-        write_specular(p, vec4(sum_specular.rgb, sum_variance));
-
+        vec4 filtered_color = sum_color / vec4(vec3(sum_w), sum_w * sum_w);
+        vec4 filtered_spec = sum_specular / vec4(vec3(sum_w_spec), sum_w_spec * sum_w_spec);
         if (control.level == 0)
         {
-            imageStore(svgf_color_hist, p, vec4(sum_color, sum_variance));
-            imageStore(svgf_hist_specular, p, vec4(sum_specular, sum_variance));
+            imageStore(diffuse_hist, p, filtered_color);
+            imageStore(spec_hist, p, filtered_spec);
```

Comment:
Is the history data really only supposed to be stored after the first iteration instead of all iterations?